cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,804 @@
+                                                apiClient, action == Action.DELETED ? null : kafkaClusterRebalance));
+
+                                } else {
+                                    updateStatus(kafkaClusterRebalance, desiredStatus, clusterRebalanceOperator,
+                                            new InvalidResourceException("Kafka resouce lacks 'cruiseControl' declaration "
+                                                    + "': No deployed Cruise Control for doing a rebalance."));
+                                }
+                            } else {
+                                updateStatus(kafkaClusterRebalance, desiredStatus, clusterRebalanceOperator, result.cause());
+                            }
+                        });
+                    } else {
+                        updateStatus(kafkaClusterRebalance, desiredStatus, clusterRebalanceOperator,
+                                new InvalidResourceException("Resource lacks label '"
+                                        + Labels.STRIMZI_CLUSTER_LABEL
+                                        + "': No cluster related to a possible rebalance."));
+                    }
+                }
+
+                @Override
+                public void onClose(KubernetesClientException e) {
+                    if (e != null) {
+                        throw e;
+                    }
+                }
+
+            });
+            return null;
+        });
+    }
+
+    public static Future<KafkaClusterRebalance> updateStatus(KafkaClusterRebalance clusterRebalance,
+                                                             KafkaClusterRebalanceStatus desiredStatus,
+                                                             CrdOperator<KubernetesClient, KafkaClusterRebalance, KafkaClusterRebalanceList, DoneableKafkaClusterRebalance> clusterRebalanceOperations,
+                                                             Throwable e) {
+        if (e != null) {
+            StatusUtils.setStatusConditionAndObservedGeneration(clusterRebalance, desiredStatus, e);
+        } else if (desiredStatus.getConditions() != null && desiredStatus.getConditions().get(0).getType() != null) {
+            StatusUtils.setStatusConditionAndObservedGeneration(clusterRebalance, desiredStatus,
+                    desiredStatus.getConditions().get(0).getType());
+        } else {
+            throw new IllegalArgumentException("Status related exception and type cannot be both null");
+        }
+        StatusDiff diff = new StatusDiff(clusterRebalance.getStatus(), desiredStatus);
+        if (!diff.isEmpty()) {
+            KafkaClusterRebalance copy = new KafkaClusterRebalanceBuilder(clusterRebalance).build();
+            copy.setStatus(desiredStatus);
+            return clusterRebalanceOperations.updateStatusAsync(copy);
+        }
+        return Future.succeededFuture(clusterRebalance);
+    }
+
+    enum State {
```

Comment:
but arent't the api supposed to be used somehow by users as well? we provide them even on Maven right? so a user can develop its own application interacting with Strimzi using this API. In this case we are talking about internal stuff related to the rebalance operator state machine.

Reply:
Yes, but the nature of the contract is different. If, in the `api` and also in the CRD JSON Schema it's an enum then users are entitled to conclude that items won't be added to the enumeration (e.g. users of the `api` module can write a `switch` over the enum with cases for each value). If we later need to add a value we're breaking that code. If it's just a String then they shouldn't be making those kinds of assumptions.

Reply:
So I am confused, I was suggesting to leave this enum here. It seems you are on the same page @tombentley ?

Reply:
I think it's OK here, yes.

Reply:
So if this state is internal only, it is not used in the custom resources then? My understanding was that this is in the Status of the CR and I expected that users are interested in this information -> E.g. the proposal is pending, I should wait more, the proposal is ready, I should review it. That seems like exactly the thing which the user wants to do using the `api`. If we pretend that the status in the CR is just opaque String, it will force the user to write their own Enum and they will have to for every release dig the possible states out of the docs and they will have no build level checks that the old states still exist. So if this is part of the status, it should be AFAIC part of the `api`.

Reply:
The states should be public and documented, but we shouldn't imply that the ones which exist today will be the only ones. It's the exhaustivity which is the problem, not the visibility.

Reply:
I'm not sure I really understand it ... it is public APi, users need to understand them and use them, but we make their life miserable by forcing the to dig it out of the docs and double check it every release?

On a practical level, the list of the states in docs is as exhaustive and final as Enum would be. It is just much harder to use.